msgpack/src/main/scala/fs2/data/msgpack/low/internal/ItemParser.scala

```diff
@@ -0,0 +1,114 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fs2
+package data
+package msgpack
+package low
+package internal
+
+import fs2.data.msgpack.low.internal.Helpers._
+import fs2.data.msgpack.low.internal.FormatParsers._
+import scodec.bits._
+
+private[low] object ItemParser {
+
+  /** Parses a single item from the stream.
+    */
+  private def parseItem[F[_]](ctx: ParserContext[F])(implicit
+      F: RaiseThrowable[F]): Pull[F, MsgpackItem, ParserContext[F]] = {
+    requireOneByte(ctx) flatMap { res => {
+      val byte = res.result
+      val ctx = res.toContext
+
+      byte match {
+        case Headers.Nil       => Pull.pure(ctx.prepend(MsgpackItem.Nil))
+        case Headers.NeverUsed => Pull.raiseError(new MsgpackParsingException("Reserved value 0xc1 used"))
+        case Headers.False     => Pull.pure(ctx.prepend(MsgpackItem.False))
+        case Headers.True      => Pull.pure(ctx.prepend(MsgpackItem.True))
+        case Headers.Bin8      => parseBin(1, ctx)
+        case Headers.Bin16     => parseBin(2, ctx)
+        case Headers.Bin32     => parseBin(4, ctx)
+        case Headers.Ext8      => parsePlainExt(1, ctx)
+        case Headers.Ext16     => parsePlainExt(2, ctx)
+        case Headers.Ext32     => parsePlainExt(4, ctx)
+        case Headers.Float32   => parseSimpleType(MsgpackItem.Float32)(4, ctx)
+        case Headers.Float64   => parseSimpleType(MsgpackItem.Float64)(8, ctx)
+        case Headers.Uint8     => parseSimpleType(MsgpackItem.UnsignedInt)(1, ctx)
+        case Headers.Uint16    => parseSimpleType(MsgpackItem.UnsignedInt)(2, ctx)
+        case Headers.Uint32    => parseSimpleType(MsgpackItem.UnsignedInt)(4, ctx)
+        case Headers.Uint64    => parseSimpleType(MsgpackItem.UnsignedInt)(8, ctx)
+        case Headers.Int8      => parseSimpleType(MsgpackItem.SignedInt)(1, ctx)
+        case Headers.Int16     => parseSimpleType(MsgpackItem.SignedInt)(2, ctx)
+        case Headers.Int32     => parseSimpleType(MsgpackItem.SignedInt)(4, ctx)
+        case Headers.Int64     => parseSimpleType(MsgpackItem.SignedInt)(8, ctx)
+        case Headers.FixExt1   => parseFixExt(1, ctx)
+        case Headers.FixExt2   => parseFixExt(2, ctx)
+        case Headers.FixExt4   => parseFixExt(4, ctx)
+        case Headers.FixExt8   => parseFixExt(8, ctx)
+        case Headers.Str8      => parseString(1, ctx)
+        case Headers.Str16     => parseString(2, ctx)
+        case Headers.Str32     => parseString(4, ctx)
+        case Headers.Array16   => parseArray(2, ctx)
+        case Headers.Array32   => parseArray(4, ctx)
+        case Headers.Map16     => parseMap(2, ctx)
+        case Headers.Map32     => parseMap(4, ctx)
+        case _ => {
+          // Positive fixint
+          if ((byte & 0x80) == 0) {
```

Comment:
If it doesn't interfere with the suggestion about `@switch` above, I think this would be more readable if the if-else construct was inlined in to the pattern matching, avoiding the nesting.